File: backend/src/main/java/com/eitek/incidents/application/services/IncidentService.java
package com.eitek.incidents.application.services;

import lombok.RequiredArgsConstructor;

import java.time.LocalDate;
import java.util.List;

import org.springframework.stereotype.Service;

import com.eitek.incidents.application.dto.AssigneeDTO;
import com.eitek.incidents.application.dto.IncidentDTO;
import com.eitek.incidents.domain.exceptions.InsufficientPermissionsException;
import com.eitek.incidents.domain.exceptions.ResourceNotFoundException;
import com.eitek.incidents.domain.models.Assignee;
import com.eitek.incidents.domain.models.Incident;
import com.eitek.incidents.domain.models.IncidentStatus;
import com.eitek.incidents.domain.repositories.AssigneeRepository;
import com.eitek.incidents.domain.repositories.IncidentRepository;
import com.eitek.incidents.domain.services.IncidentAssignmentValidator;

@Service
@RequiredArgsConstructor
public class IncidentService {
    private final IncidentRepository incidentRepository;
    private final AssigneeRepository assigneeRepository;
    private final IncidentAssignmentValidator assignmentValidator = new IncidentAssignmentValidator();

    public IncidentDTO createIncident(IncidentDTO incidentDTO) {
        Incident incident = Incident.builder()
                .title(incidentDTO.getTitle())
                .status(incidentDTO.getStatus() != null ? incidentDTO.getStatus() : IncidentStatus.OPEN)
                .description(incidentDTO.getDescription() != null ? incidentDTO.getDescription() : "")
                .build();
        Incident savedIncident = incidentRepository.save(incident);
        return mapToDTO(savedIncident);
    }

    public IncidentDTO getIncidentById(Long id) {
        Incident incident = incidentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Incident", "id", id));
        return mapToDTO(incident);
    }

    public List<IncidentDTO> getAllIncidents() {
        List<Incident> incidents = incidentRepository.findAll();
        return incidents.stream()
                .map(this::mapToDTO)
                .toList();
    }

    public IncidentDTO updateIncident(Long id, IncidentDTO incidentDTO) {
        Incident existingIncident = incidentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Incident", "id", id));

        if (incidentDTO.getTitle() != null) {
            existingIncident.setTitle(incidentDTO.getTitle());
        }
        if (incidentDTO.getStatus() != null) {
            existingIncident.setStatus(incidentDTO.getStatus());
        }
        if (incidentDTO.getDescription() != null) {
            existingIncident.setDescription(incidentDTO.getDescription());
        }
        
        // Actualizar updatedAt
        existingIncident.setUpdatedAt(java.time.LocalDateTime.now());
        
        // Si se está asignando a alguien, validar permisos
        if (incidentDTO.getAssigneeId() != null) {
            assignIncident(existingIncident, incidentDTO.getAssigneeId(), incidentDTO.getAssignerId());
        }

        Incident updatedIncident = incidentRepository.save(existingIncident);
        return mapToDTO(updatedIncident);
    }
    
    /**
     * Asigna una incidencia validando los permisos del assignee
     * @param incident La incidencia a asignar
     * @param targetAssigneeId ID del assignee al que se asigna
     * @param assignerId ID del assignee que hace la asignación (puede ser null para auto-asignación)
     */
    private void assignIncident(Incident incident, Long targetAssigneeId, Long assignerId) {
        Assignee targetAssignee = assigneeRepository.findById(targetAssigneeId)
                .orElseThrow(() -> new ResourceNotFoundException("Assignee", "id", targetAssigneeId));
        
        // Auto-asignación
        if (assignerId == null || assignerId.equals(targetAssigneeId)) {
            if (!assignmentValidator.canSelfAssign(targetAssignee, incident)) {
                throw new InsufficientPermissionsException("No puede auto-asignarse esta incidencia");
            }
        } else {
            // Asignación a otro
            Assignee assigner = assigneeRepository.findById(assignerId)
                    .orElseThrow(() -> new ResourceNotFoundException("Assignee", "id", assignerId));
            
            if (!assigner.isAdmin()) {
                throw new InsufficientPermissionsException(assigner.getRole().name(), "asignar incidentes");
            }
        }
        
        incident.setAssigneeId(targetAssigneeId);
        incident.setDateAssigned(LocalDate.now());
    }

    public void deleteIncident(Long id, Long deleterId) {
        // Verificar que la incidencia existe
        incidentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Incident", "id", id));
        
        // Validar permisos de eliminación
        if (deleterId == null) {
            throw new InsufficientPermissionsException("Se requiere autenticación para eliminar incidentes");
        }
        
        Assignee deleter = assigneeRepository.findById(deleterId)
                .orElseThrow(() -> new ResourceNotFoundException("Assignee", "id", deleterId));
        
        if (!deleter.isAdmin()) {
            throw new InsufficientPermissionsException(deleter.getRole().name(), "eliminar incidentes");
        }
        
        incidentRepository.deleteById(id);
    }

    public IncidentDTO mapToDTO(Incident incident) {
        AssigneeDTO assigneeDTO = null;
        
        if (incident.getAssigneeId() != null) {
            assigneeDTO = assigneeRepository.findById(incident.getAssigneeId())
                .map(assignee -> AssigneeDTO.builder()
                    .id(assignee.getId())
                    .name(assignee.getName())
                    .role(assignee.getRole())
                    .isActive(assignee.isActive())
                    .createdAt(assignee.getCreatedAt())
                    .updatedAt(assignee.getUpdatedAt())
                    .build())
                .orElse(null);
        }
        
        return IncidentDTO.builder()
                .id(incident.getId())
                .title(incident.getTitle())
                .status(incident.getStatus())
                .description(incident.getDescription())
                .assigneeId(incident.getAssigneeId())
                .assignee(assigneeDTO)
                .dateAssigned(incident.getDateAssigned())
                .createdAt(incident.getCreatedAt())
                .updatedAt(incident.getUpdatedAt())
                .build();        
    }
}

File: frontend/src/application/services/IncidentService.ts
import type { IncidentDTO } from "../dto/IncidentDTO";
import { Incident } from "../../core/domain/Incidents";
import { Assignee } from "../../core/domain/Assignee";
import { AssigneeRoles } from "../../core/domain/AssigneeRoles";
import { IncidentRepository } from "../../infrastructure/repositories/IncidentRepository";

export class IncidentService {
    private repository: IncidentRepository;

    constructor() {
        this.repository = new IncidentRepository();
    }

    async getAllIncidents(): Promise<Incident[]> {
        const dtos = await this.repository.findAll();
        return dtos.map(dto => this.toDomain(dto));
    }

    async getIncidentById(id: number): Promise<Incident> {
        const dto = await this.repository.findById(id);
        return this.toDomain(dto);
    }

    async createIncident(data: { 
        title: string; 
        description: string;
        status?: string;
    }): Promise<Incident> {
        const dto = await this.repository.create(data);
        return this.toDomain(dto);
    }

    async updateIncident(
        id: number, 
        data: { 
            title?: string; 
            description?: string;
            status?: string;
            assigneeId?: number;
            assignerId?: number;
        }
    ): Promise<Incident> {
        const dto = await this.repository.update(id, data);
        return this.toDomain(dto);
    }

    async deleteIncident(id: number, userId: number): Promise<void> {
        await this.repository.delete(id, userId);
    }

    private toDomain(dto: IncidentDTO): Incident {
        return new Incident({
            id: dto.id,
            title: dto.title,
            description: dto.description,
            assigneeId: dto.assigneeId,
            assignee: dto.assignee ? new Assignee({
                id: dto.assignee.id,
                name: dto.assignee.name,
                role: dto.assignee.role as AssigneeRoles,
                isActive: dto.assignee.isActive,
                createdAt: new Date(dto.assignee.createdAt),
                updatedAt: new Date(dto.assignee.updatedAt)
            }) : null,
            status: dto.status as any,
            createdAt: new Date(dto.createdAt),
            updatedAt: new Date(dto.updatedAt)
        });
    }
}
